internal/controller/object/object.go

```diff
@@ -340,3 +357,177 @@ func setObserved(obj *v1alpha1.Object, observed *unstructured.Unstructured) erro
+				return errors.Wrap(err, errPatchFromReferencedResource)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *external) isNotFound(obj *v1alpha1.Object, err error) bool {
+	isNotFound := false
+
+	if kerrors.IsNotFound(err) {
+		isNotFound = true
+	} else if meta.WasDeleted(obj) {
+		// If the Object resource was being deleted but the external resource is
+		// not deletable as management policy is specified, we should return the
+		// external resource not found, so that Object can be deleted by managed
+		// resource reconciler. Otherwise, the reconciler will try to delete the
+		// external resource which breaks the management policy.
+		if !obj.Spec.ManagementPolicy.IsActionAllowed(v1alpha1.ObjectActionDelete) {
+			c.logger.Debug("Managed resource was deleted but external resource is undeletable.")
+			isNotFound = true
+		}
+	}
+
+	return isNotFound
+}
+
+func (c *external) handleLastApplied(obj *v1alpha1.Object, last, desired *unstructured.Unstructured) (managed.ExternalObservation, error) {
+	isUpToDate := false
+
+	if !obj.Spec.ManagementPolicy.IsActionAllowed(v1alpha1.ObjectActionUpdate) {
+		// Treated as up-to-date to skip last applied annotation update since we
+		// do not create or update the external resource.
+		isUpToDate = true
+	} else if last != nil && equality.Semantic.DeepEqual(last, desired) {
+		// Mark as up-to-date since last is equal to desired
+		isUpToDate = true
+	}
+
+	if isUpToDate {
+		c.logger.Debug("Up to date!")
+
+		obj.Status.SetConditions(xpv1.Available())
+
+		return managed.ExternalObservation{
+			ResourceExists:   true,
+			ResourceUpToDate: true,
+		}, nil
+	}
+
+	return managed.ExternalObservation{
+		ResourceExists:   true,
+		ResourceUpToDate: false,
+	}, nil
+}
+
+type objFinalizer struct {
+	resource.Finalizer
+	client client.Client
+}
+
+type refFinalizerFn func(context.Context, *unstructured.Unstructured, string) error
+
+func (f *objFinalizer) handleRefFinalizer(ctx context.Context, obj *v1alpha1.Object, finalizerFn refFinalizerFn) error {
+	// Loop through references to resolve each referenced resource
+	for _, ref := range obj.Spec.References {
+		res := &unstructured.Unstructured{}
+		res.SetAPIVersion(ref.FromObject.APIVersion)
+		res.SetKind(ref.FromObject.Kind)
+		// Try to get referenced resource
+		err := f.client.Get(ctx, client.ObjectKey{
+			Namespace: ref.FromObject.Namespace,
+			Name:      ref.FromObject.Name,
+		}, res)
+
+		if err != nil {
+			return errors.Wrap(err, errGetReferencedResource)
+		}
+
+		finalizerName := refFinalizerNamePrefix + string(obj.UID)
+		if err = finalizerFn(ctx, res, finalizerName); err != nil {
+			return err
+		}
+	}
+
+	return nil
+
+}
+
+func (f *objFinalizer) AddFinalizer(ctx context.Context, res resource.Object) error {
+	obj, ok := res.(*v1alpha1.Object)
+	if !ok {
+		return errors.New(errNotKubernetesObject)
+	}
+
+	if meta.FinalizerExists(obj, objFinalizerName) {
+		return nil
+	}
+	meta.AddFinalizer(obj, objFinalizerName)
```

Comment:
This is because, when we call `WithFinalizer` in `Setup()` to register the custom behavior on how to add and remove finalizers to and from the managed resource:

```go
managed.WithFinalizer(&objFinalizer{client: mgr.GetClient()})
```

It will replace the default behavior defined in crossplane-runtime (pkg/resource/api.go):

```go
// AddFinalizer to the supplied Managed resource.
func (a *APIFinalizer) AddFinalizer(ctx context.Context, obj Object) error {
	if meta.FinalizerExists(obj, a.finalizer) {
		return nil
	}
	meta.AddFinalizer(obj, a.finalizer)
	return errors.Wrap(a.client.Update(ctx, obj), errUpdateObject)
}
```

I know it's not ideal, but that appears to be the best option I can see :-) Any suggestion?